### PR TITLE
Take SuppressGCTransition into account in IL stub caching

### LINKED ISSUE #46184: UnmanagedCallersOnly fails in Debug mode, with SuppressGCTransition and certain signatures
### Description

The following code results in `Fatal error. Invalid Program: attempted to call a UnmanagedCallersOnly method from managed code.`.

```cs
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

unsafe class Program
{
    [DllImport("NativeDll", CallingConvention = CallingConvention.Cdecl)]
    public static extern void set_callback(delegate* unmanaged[Cdecl]<int, void> callback);

    [DllImport("NativeDll", CallingConvention = CallingConvention.Cdecl), SuppressGCTransition]
    public static extern void set_value(void* unused, int value);

    [DllImport("NativeDll", CallingConvention = CallingConvention.Cdecl)]
    public static extern void call_callback(void* unused, int unused2);

    [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvCdecl) })]
    private static void PrintInt(int value)
    {
        Console.WriteLine(value);
        Console.ReadKey(true);
    }

    static void Main()
    {
        set_callback(&PrintInt);
        set_value(null, 1234);
        call_callback(null, 1234);  // crashes here!
    }
}
```

```cpp
void (*globalCallback)(int value) = nullptr;
int globalValue = 0;

extern "C" __declspec(dllexport) void set_callback(void (*callback)(int value)) {
	globalCallback = callback;
}

extern "C" __declspec(dllexport) void set_value(void*, int value) {
	globalValue = value;
}

extern "C" __declspec(dllexport) void call_callback(void*, int) {
	globalCallback(globalValue);
}
```

### Configuration

.NET Version: 5.0.101
OS: Windows 10 (18363.1256)
Architecture: x64

### Regression?

N/A, since a lot of this stuff didn't even exist on 3.1.

### Other information

Notably, if at least one of the following is true, then the crash no longer occurs:

1. x86 is used instead of x64.
2. `SuppressGCTransition` is removed from `set_value`.
3. `set_value` is modified to take a `long` instead of an `int`.
4. `call_callback` is modified to take a `long` instead of an `int`.
5. The code is run under Release mode.

### PR DESCRIPTION
**Pull request by elinor-fung**
Use the updated stub flags that have `SuppressGCTransition`  taken into account when creating the hash blob used for caching IL stubs.

Fix #46184

cc @AaronRobinsonMSFT @jkoritzinsky

### CODE REVIEW

File: src/coreclr/vm/ilstubcache.cpp [hunk 1]
```diff
@@ -501,57 +484,38 @@ MethodDesc* ILStubCache::GetStubMethodDesc(
...
-
-                    phe->m_pMethodDesc   = pMD;
-                    pBlob = pBlobHolder;
-                    phe->m_pBlob         = pBlob;
+                pBlobHolder.SuppressRelease();
+                pBlob = pBlobHolder;
 
-                    _ASSERTE(pParams->m_cbSizeOfBlob == cbSizeOfBlob);
-                    memcpy(pBlob, pParams, cbSizeOfBlob);
+                phe = new ILStubCacheEntry{ pMD, pBlob };
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/ilstubcache.cpp:498]: This is a throwing `new` - is that okay in this code path?

File: src/tests/Interop/PInvoke/Attributes/SuppressGCTransition/SuppressGCTransitionNative.cpp [hunk 2]
```diff
@@ -21,3 +21,15 @@ BOOL DLL_EXPORT STDMETHODVCALLTYPE NextUInt(/* out */ uint32_t *n)
     *n = (++_n);
     return TRUE;
 }
+
+typedef int (STDMETHODVCALLTYPE *CALLBACKPROC)(int n);
+
+extern "C"
+BOOL DLL_EXPORT InvokeCallback(CALLBACKPROC cb, int* n)
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/PInvoke/Attributes/SuppressGCTransition/SuppressGCTransitionNative.cpp:28]: Please add the calling convention here.

File: src/coreclr/vm/ilstubcache.cpp [hunk 3]
```diff
@@ -501,57 +484,38 @@ MethodDesc* ILStubCache::GetStubMethodDesc(
...
+            if (phe == NULL)
             {
-                if (bNew)
-                {
-                    pBlobHolder.SuppressRelease();
-
-                    phe->m_pMethodDesc   = pMD;
-                    pBlob = pBlobHolder;
-                    phe->m_pBlob         = pBlob;
+                pBlobHolder.SuppressRelease();
```
Comment (jkotas) [inline src/coreclr/vm/ilstubcache.cpp:495]: This should be done only after the entry was successfully added.

File: src/coreclr/vm/ilstubcache.cpp [hunk 4]
```diff
@@ -562,7 +526,7 @@ MethodDesc* ILStubCache::GetStubMethodDesc(
     RETURN pMD;
 }
 
-void ILStubCache::DeleteEntry(void* pParams)
+void ILStubCache::DeleteEntry(ILStubHashBlob* pHashBlob)
```
Comment (jkotas) [inline src/coreclr/vm/ilstubcache.cpp:529]: Is this used? Where is it called from?

File: src/coreclr/vm/ilstubcache.cpp (hunk 1 above)
Comment (jkotas) [reply src/coreclr/vm/ilstubcache.cpp:498]: It should be ok - `m_hashMap.Add` is throwing too.

File: src/coreclr/vm/ilstubcache.h [hunk 5]
```diff
@@ -112,49 +91,29 @@ class ILStubCache : private CClosedHashBase
...
-    virtual void* GetKey(                   // The data to hash on.
-        BYTE*        pElement);             // The element to return data ptr for.
+private: // Inner classes
+    struct ILStubCacheEntry
+    {
+        MethodDesc *m_pMethodDesc;
+        ILStubHashBlob *m_pBlob;
+    };
+
+    class ILStubCacheTraits : public DeleteElementsOnDestructSHashTraits<DefaultSHashTraits<ILStubCacheEntry *>>
```
Comment (jkotas) [inline src/coreclr/vm/ilstubcache.h:101]: Small element types like `ILStubCacheEntry` are typically stored in SHash inplace for better locality and efficiency. Consider changing the element type to `ILStubCacheEntry`.

File: src/coreclr/vm/ilstubcache.cpp (hunk 4 above)
Comment (elinor-fung) [reply src/coreclr/vm/ilstubcache.cpp:527]: It is used by `NDirect::RemoveILStubCacheEntry` (when there is an error during stub generation):
https://github.com/dotnet/runtime/blob/2f2593177dafbe702407fe0b7ac156a7829b7ee6/src/coreclr/vm/dllimport.cpp#L3992-L3994

File: src/coreclr/vm/ilstubcache.cpp [hunk 6]
```diff
@@ -501,42 +484,28 @@ MethodDesc* ILStubCache::GetStubMethodDesc(
 
         if (SF_IsSharedStub(dwStubFlags))
         {
-            size_t cbSizeOfBlob = pParams->m_cbSizeOfBlob;
+            size_t cbSizeOfBlob = pHashBlob->m_cbSizeOfBlob;
             AllocMemHolder<ILStubHashBlob> pBlobHolder( m_heap->AllocMem(S_SIZE_T(cbSizeOfBlob)) );
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/ilstubcache.cpp:488]: Let's move this into the `if (phe == NULL)` branch. I don't see any reason to allocate this prior to looking up what might already exist.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by elinor-fung** [comment]
@AaronRobinsonMSFT @jkoritzinsky thoughts on if we should try to take (a version of) this for 5?

**Comment by jkoritzinsky** [comment]
I think we should. If we don't, users can end up with random GC pauses where one of their IL stubs hit a cache entry that was marked SuppressGCTransition.

**Comment by AaronRobinsonMSFT** [comment]
@elinor-fung It's not a bad idea given it isn't easy to diagnose. I would like to see as much of the testing as was added here brought down to .NET 5 though.

**Pull request merged**